File: cpp/7_reverseInteger.cpp
#include <iostream>
#include <string>
using namespace std;

// Runtime: 0ms (100%) Memory: 6.4 MB (7.17%)
int my_reverse(int x) {
    if (x == 0)
        return 0;
    string s = to_string(x);
    string res;

    // just like scan from rightside
    bool isPos = isdigit(s[0]);
    s = isPos ? s : s.substr(1, s.length() - 1);
    int l = s.length();
    bool islastconstZero = (s[s.length() - 1] == '0') ? true : false;
    int i = 0;
    while (i < l) {
        islastconstZero = (islastconstZero && s[s.length() - 1 - i] == '0') ? true : false;
        if (!islastconstZero)
            res += s[s.length() - 1 - i];
        i++;
    }
    res = isPos ? res : '-' + res;
    try {
        return stoi(res);
    }
    catch (const std::out_of_range&) {
        return 0;
    }
}

// ref: https://stackoverflow.com/a/59029832/9432281
// Runtime: 0ms(100 %) Memory : 5.7 MB(96.28 %)
int ref_reverse(int x) {
    long long int ans = 0; //should define with long long int or some compiler will return wrong value
    while (x!=0) 
    {
        ans = ans * 10 + (x % 10);
        x /= 10;
    }

    if (ans > 2147483647 || ans < -2147483648) {
        return 0;
    }
    else
    {
        return int(ans);
    }
}

int main() {
    cout << ref_reverse(1534236469) << endl;
    return 0;
}

// Use this template always on leetcode to improve runtime.
static const int _ = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 0;
}();

File: python/7_reverseInteger.py
# Runtime: 32 ms, faster than 74.91% of Python3 online submissions for Reverse Integer.
# Memory Usage: 14.2 MB, less than 43.54% of Python3 online submissions for Reverse Integer.
def my_reverse(x):
    if x == 0:
        return 0
    i = 0
    res = ""
    isPos = x > 0
    x = x if isPos else -1 * x
    str_x = str(x)
    lastConstZero = str_x[len(str_x) - 1] == '0'
    while (i < len(str_x)):
        lastConstZero = lastConstZero and str_x[len(str_x) - 1 - i] == '0'
        if not lastConstZero:
            res += str_x[len(str_x) - 1 - i]
        i += 1
    result = int(res) if isPos else -1 * int(res)
    if result > 2147483647 or result < -2147483648:
        return 0
    else:
        return result



# runtime: 32 ms(74.91%), memory: 14.2 MB(43.54%)
def ref_reverse(x):
    isPos = x > 0
    ans=0
    x = -1 * x if not isPos else x
    while(x!=0):
        ans = ans * 10 + (x % 10)
        x = int(x/10)

    res = -1 * ans if not isPos else ans
    if (res > 2147483647 or res < -2147483648):
        return 0
    else:
        return  res


if __name__ == "__main__":
    # 1534236469
    # -123
    print(reverse(-123))
    
